### DISCUSSION TITLE
How to write Rego with confidence?

### DISCUSSION THREAD

**Discussion by mbrg** [open-policy-agent/community#39] [OPA and Rego] [answered]
We find writing Rego policies to be very error-prone. We use both the [VS Code](https://github.com/open-policy-agent/vscode-opa) and [IntelliJ](https://github.com/open-policy-agent/opa-idea-plugin) extensions, but they both lack features like error-highlighting, auto-complete or IntelliSense. Most issues are small typos in a variable names or operators.

How do you write policies with confidence? Any tips would be much appreciated.

**Comment by anderseknert** [comment, marked as answer]
Hi Michael!

And thanks for bringing this up! I don't have a definitive answer, so I hope others can share their tips here as well, but here's some of my thoughts.

As the [OPA Survey 2021](https://blog.openpolicyagent.org/open-policy-agent-2021-survey-summary-e749bbd7b824) showed, a better OPA debugging experience is at the top of the wish list for the OPA community at large. Improving that is, and will continue to be, a high priority for OPA development, and we'll see some improvements—like a new print function for debugging—in the next OPA release, with more improvements in the type of tooling you mention planned for later this year. 

Some things that I think would help the Rego authoring and debugging experience:

* A proper [debugger](https://github.com/open-policy-agent/opa/issues/3191)
* A Rego [linter](https://github.com/open-policy-agent/opa/issues/3218)
* Better description of [test assertion failures](https://github.com/open-policy-agent/opa/issues/2546)
* Improved editor support

With that said, some things that are currently there already and which I feel aren't utilized fully by many:

* Unit testing - this one is IMO the number one way to build confidence in policies - not just at the time of authoring but over time. The kind of issues you mentioned with typos should be discovered quickly with tests in place.
* OPA eval. Learning how to use OPA eval to quickly query for the value of some rule or variable given some provided input makes for a pretty nice debugging experience.

The new print function, and its ability to print even undefined variables and values, will also help a lot in scenarios where you just want to quickly know what's going on inside of a rule or comprehension. While it sounds pretty simple I've been running with the dev build of OPA for a week or two, and it's really _very_ nice to work with. Hope others will find it as useful as I do!

**Reply by mbrg** [reply]
Thanks for the quick reply!
I'm glad to hear that you are working on solutions! I've subscribed to updates from all relevant issues. We are happy to contribute as well if relevant.

To your point on unit tests, we use Rego mainly through `conftest` and couple every policy with relevant tests evaluted with `conftest verify`. However, it's just as easy to have a typo hidden in the test as it is in the policy. It's also very easy to make a mistake in the test's _input_ or _context_  because these are unstructured.

Are there any plans to support auto-complete / IntelliSense-type features?

**Reply by anderseknert** [reply]
I think so yes, but let's make sure we have proper tracking on that. Unless they're already there, could you please file issues for what you want in the respective repos?

* https://github.com/open-policy-agent/vscode-opa
* https://github.com/open-policy-agent/opa-idea-plugin/issues

These would be valuable for setting our priorities later, and will give others a chance to have a say. 

Thanks!

**Reply by mbrg** [reply]
Sure! will do. Thanks again.
